cpp/src/lists/contains.cu

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/logical.h>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/lists/contains.hpp>
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/exec_policy.hpp>
+#include <type_traits>
+
+namespace cudf {
+namespace lists {
+
+namespace {
+
+template <bool search_keys_have_nulls,
+          bool search_keys_are_all_nulls,
```

Comment:
I don't see why these need to be template parameters. `search_keys_are_all_nulls` can definitely be a function parameter, and I think both could be. I find this a bit smelly.

Reply:
I also find it weird to call it `search_keys_are_all_nulls` when it is only true when there is a single search key that is null.

Reply:
> I don't see why these need to be template parameters.

`search_keys_have_nulls` is a compile-time constant so that the check at [line#65](https://github.com/mythrocks/cudf/blob/6ebed0855dcc4e0b7a50f41eb84153bce2f11634/cpp/src/lists/contains.cu#L65) can be optimized away if there are no nulls in `search_keys`:
```c++
[d_lists, search_key_iter] __device__(auto const& row_index) {
// Bail, if search_key row is null.
if (search_keys_have_nulls && !search_key_iter[row_index]) {
  return false;
}
...
```

`search_keys_are_all_nulls` could certainly be a function parameter. I put them together since they were both null flags on `search_keys`. The invocation looked clumsy to my eye when these flags were split apart.

Reply:
> I also find it weird to call it `search_keys_are_all_nulls` when it is only true when there is a single search key that is null.

This is an artifact of having to fold the implementations of `construct_null_mask()` for `column_view` and `scalar` into a single one.
The `scalar` version would bail early, and return `ALL_NULLS` when the `search_key` is null.
That code would only be applicable in the `column_view` if all the `search_keys` rows were null. Hence, the variable name.

I don't call it with `search_keys_are_all_null = true` because I wasn't sure this might be worth incurring the cost of the check.

Reply:
Fine. `search_key_is_null_scalar` might be a better name though.